src/sizes/retained_size_action.cpp

```diff
@@ -126,16 +126,17 @@ jint JNICALL tagObjectOfTaggedClass(jlong classTag, jlong size, jlong *tagPtr, j
 
 jvmtiError walkHeapFromObjects(jvmtiEnv *jvmti,
                                const std::vector<jobject> &objects,
-                               const std::chrono::steady_clock::time_point &finishTime) {
+                               const std::chrono::steady_clock::time_point &finishTime,
+                               const std::string &cancellationFileName) {
     jvmtiError err = JVMTI_ERROR_NONE;
     if (!objects.empty()) {
         jvmtiHeapCallbacks cb;
         std::memset(&cb, 0, sizeof(jvmtiHeapCallbacks));
         cb.heap_reference_callback = reinterpret_cast<jvmtiHeapReferenceCallback>(&spreadInfo);
         int heapWalksCnt = 0;
         for (auto &object : objects) {
-            if (finishTime < std::chrono::steady_clock::now()) {
-                return MEMORY_AGENT_TIMEOUT_ERROR;
+            if (shouldStopExecution(finishTime, cancellationFileName)) {
+                return MEMORY_AGENT_INTERRUPTED_ERROR;
```

Comment:
Shouldn't we distinguish timeout and cancel by file?

Reply:
I think we should look at this check as an interruption point. From this point of view, there are two situations when the agent can be interrupted: timeout and cancellation by file. Separating these two will force the user to write two condition checks in a row, while he just wants to check if the agent was interrupted somehow.